src/main/scala/calc/Compiler.scala

```diff
@@ -62,21 +61,212 @@ object Compiler {
+
+        // Do f__ref.func = f
+        val objectRef = irt.VarRef(objectName)(irtpe.AnyType)
+        val dotSelect = irt.JSDotSelect(objectRef, irt.Ident("func"))
+        val recursiveCapture = Some(objectName.name, objectRef)
+
+        // Substitute f with f__ref.func in closure body
+        val newBody = substitute(tree, t.name.name, new SelectT(objectName.name, "func") { val tpe = tree.tpe })
+        val assign = irt.Assign(dotSelect, closure(newBody.asInstanceOf[ClosureT], recursiveCapture))
+
+        // Do the binding, i.e f = f__ref.func
+        val attrAccess = irt.JSDotSelect(objectRef, irt.Ident("func"))
+        val binding = irt.VarDef(irt.Ident(bindingName), t.value.tpe.irtype, false, attrAccess)
+        irt.Block(List(objectDef, assign, binding, body))
+      case _ => {
+        val bindingRhs = expr(t.value)
+        val binding = irt.VarDef(irt.Ident(t.name.name), t.value.tpe.irtype, false, bindingRhs)
+        irt.Block(List(binding, body))
+      }
+    }
+  }
+
+  def ifElse(t: IfT) = { implicit val pos = t.pos
+    val cond = irt.BinaryOp(irt.BinaryOp.!==, expr(t.cond), irt.DoubleLiteral(0))
+    val thenp = expr(t.thenp)
+    val elsep = expr(t.elsep)
+    irt.If(cond, thenp, elsep)(t.tpe.irtype)
+  }
+
+  def closure(t: ClosureT, recursiveCapture: Option[(String, irt.VarRef)] = None) = { implicit val pos = t.pos
+    val paramBoxed = t.params map { ident =>
+      val mangled = irt.Ident(mangle(ident.name))
+      irt.ParamDef(mangled, irtpe.AnyType, false, false)
+    }
+
+    val freeVars = freeVariables(t)
+    val captureParamsRef = freeVars.map({ v => irt.VarRef(irt.Ident(v._1))(v._2.irtype) }).toList
+    val recursiveCaptureRef = (recursiveCapture map { kv => kv._2 }).toList
+
+    val captureParamsDef = freeVars.map({ v => irt.ParamDef(irt.Ident(v._1), v._2.irtype, false, false) }).toList
+    val recursiveCaptureDef = (recursiveCapture map { kv => irt.ParamDef(irt.Ident(kv._1), irtpe.AnyType, false, false) }).toList
+
+    val paramUnbox = t.params map { ident =>
+      val mangled = irt.Ident(mangle(ident.name))
+      val varRef = irt.VarRef(mangled)(irtpe.AnyType)
+      val varUnbox = irt.Unbox(varRef, 'D')
+      irt.VarDef(irt.Ident(ident.name), irtpe.DoubleType, false, varUnbox)
+    }
+    val body = expr(t.body)
+    val block = irt.Block(paramUnbox ++ List(body))
+
+    irt.Closure(captureParamsDef ++ recursiveCaptureDef, paramBoxed, block,
+      captureParamsRef ++ recursiveCaptureRef)
+  }
+
+  def call(t: CallT) = { implicit val pos = t.pos
+    val args = t.args map expr
+    irt.Unbox(irt.JSFunctionApply(expr(t.fun), args), 'D')
+  }
+
+  def foreignCall(t: ForeignCallT) = { implicit val pos = t.pos
+    val receiver = irt.LoadModule(irtpe.ClassType(encodeClassName(t.clsName + "$")))
+    val method = irt.Ident(t.methodName)
+    val args = t.args map expr
+    irt.Unbox(irt.Apply(receiver, method, args)(t.tpe.irtype), 'D')
```

Comment:
That was probably there before, but I notice now that the `Unbox` is not needed for an `Apply` node. That is because "foreign" calls, as you call them, are actually the *normal* calls of Scala.js, which are statically typed. `JSFunctionApply` nodes are the "foreign" calls of the Scala.js IR, as they call JavaScript functions, which are dynamically typed.